routing/osrm_helpers.cpp

```diff
@@ -0,0 +1,300 @@
+       {
+         return (r1.m_dist < r2.m_dist);
+       });
+
+  size_t const n = min(m_candidates.size(), maxCount);
+  for (size_t j = 0; j < n; ++j)
+  {
+    OsrmMappingTypes::FtSeg & seg = segments[j];
+    Candidate const & c = m_candidates[j];
+
+    seg.m_fid = c.m_fid;
+    seg.m_pointStart = c.m_segIdx;
+    seg.m_pointEnd = c.m_segIdx + 1;
+
+    segmentSet.insert(&seg);
+  }
+
+  OsrmFtSegMapping::OsrmNodesT nodes;
+  m_routingMapping.m_segMapping.GetOsrmNodes(segmentSet, nodes);
+
+  res.clear();
+  res.resize(maxCount);
+
+  for (size_t j = 0; j < maxCount; ++j)
+  {
+    size_t const idx = j;
+
+    if (!segments[idx].IsValid())
+      continue;
+
+    auto it = nodes.find(segments[idx].Store());
+    if (it == nodes.end())
+      continue;
+
+    FeatureGraphNode & node = res[idx];
+
+    if (!m_direction.IsAlmostZero())
+    {
+      // Filter income nodes by direction mode
+      OsrmMappingTypes::FtSeg const & node_seg = segments[idx];
+      FeatureType feature;
+      Index::FeaturesLoaderGuard loader(*m_pIndex, m_routingMapping.GetMwmId());
+      loader.GetFeatureByIndex(node_seg.m_fid, feature);
+      feature.ParseGeometry(FeatureType::BEST_GEOMETRY);
+      m2::PointD const featureDirection = feature.GetPoint(node_seg.m_pointEnd) - feature.GetPoint(node_seg.m_pointStart);
+      bool const sameDirection = (m2::DotProduct(featureDirection, m_direction) / (featureDirection.Length() * m_direction.Length()) > 0);
+      if (sameDirection)
+      {
+        node.node.forward_node_id = it->second.first;
+        node.node.reverse_node_id = INVALID_NODE_ID;
+      }
+      else
+      {
+        node.node.forward_node_id = INVALID_NODE_ID;
+        node.node.reverse_node_id = it->second.second;
+      }
+    }
+    else
+    {
+      node.node.forward_node_id = it->second.first;
+      node.node.reverse_node_id = it->second.second;
+    }
+
+    node.segment = segments[idx];
+    node.segmentPoint = m_candidates[j].m_point;
+    node.mwmName = mwmName;
+
+    CalculateOffsets(node);
+  }
+  res.erase(remove_if(res.begin(), res.end(),
+                      [](FeatureGraphNode const & f)
+                      {
+                        return f.mwmName.empty();
+                      }),
+            res.end());
+}
+
+void Point2PhantomNode::CalculateOffsets(FeatureGraphNode & node) const
+{
+  CalculateWeight(node.segment, node.segmentPoint, node.node.forward_node_id,
+                  node.node.forward_weight, true /* weightFromLeft */);
+  CalculateWeight(node.segment, node.segmentPoint, node.node.reverse_node_id,
+                  node.node.reverse_weight, false /* weightFromLeft */);
+
+  // need to initialize weights for correct work of PhantomNode::GetForwardWeightPlusOffset
+  // and PhantomNode::GetReverseWeightPlusOffset
+  node.node.forward_offset = 0;
+  node.node.reverse_offset = 0;
+}
+
+void Point2Node::operator()(FeatureType const & ft)
+{
+  static CarModel const carModel;
```

Comment:
What about a ```CarModel const & model = GetCarModel();``` approach? In this way we'll not have many static instances of the exactly same class all around in the code.